Revised shortest path algorithms

d-michail opened the pull request:
This is a pull-request related to issue #294. 

Main changes: 
- Added interface for shortest path algorithms
  - The interface supports SingleSourceSingleSink, SingleSource shortest path queries
  - Tried to keep the naming consistent with traditional terminology in graph-theory 
- Added k-shortest paths interface
- Bundled all shortest path related code into a shortestpath package and deprecated old classes.
- Refactored FloydWarshall
  - Added support for multi-graphs. 
  - Added separate tests for FloydWarshall support of multigraphs 
  - Fixed diameter method to return POSITIVE_INFINITY when a graph is disconnected
- Refactored Dijkstra
  - Added extra method for returning single-source shortest paths, using the traditional representation of a shortest path tree
- Minor fixes in corner-cases when returning paths
  - Now the default vertex list implementation of GraphPath supports zero edge paths 
  - Adjusted KShortestPaths to return empty list instead of null if no paths exist
  - Changed algorithms to return zero edge paths with 0 weight if the source equals the target

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/GraphPath.java:
```diff
@@ -92,8 +92,14 @@
 
-        if (edgeList.isEmpty())
-            return Collections.emptyList();
+        if (edgeList.isEmpty()) {
```
Great thanks! I literally wrote the same fix yesterday :)

d-michail commented:
Additions to the pull-request containing the following: 

 - Added an implementation of Goldberg's [ALT admissible heuristic](https://www.microsoft.com/en-us/research/publication/computing-the-shortest-path-a-search-meets-graph-theory/)  for A*. This is a really nice graph-theoretic lower-bounding technique based on landmarks and the triangle inequality.
 - Added a lightweight (package-private) implementation of the `ClosestFirstIterator` which does not support cross components traversal or listeners. This speeds up the implementation of Dijkstra significantly.
 - Added benchmark called `DijkstraShortestPathPerformanceTest` which executes all-pairs shortest paths using the different non-negative shortest paths algorithms
   - Dijkstra with ClosestFirstIterator
   - Dijkstra with the new lightweight iterator
   - A* with no heuristic (zero function)
   - A* with ALT heuristic and 1 random landmark
   - A* with ALT heuristic and 5 random landmark
   - Bidirectional Dijkstra
 - Added tests for all the above.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/FloydWarshallShortestPaths.java:
```diff
@@ -0,0 +1,437 @@
+     * {@inheritDoc}
+     */
+    @Override
+    public SingleSourcePaths<V, E> getPaths(V source)
```
This doesn't seem ideal. This method transforms the internal datastructure of FloydWarshall (its distance matrix) to the datastructure of TreeSingleSourcePaths (a map with predecessors) which is pure overhead. These data structures encode the same, but in a different format. Whenever the user queries a path from TreeSingleSourcePaths, the getPath(V sink) method in TreeSingleSourcePaths performs the same steps as the getPath(V source, V sink) in FloydWarshall would do. Consequently, if the user wants to query the paths from a single source u to various sink vertices v, the user would be better off to repeatedly invoke the getPath(V u, V v) method than the getPath(V v) method.

For this particular class, it would be better if it would directly implement SingleSourcePaths. An invocation to getPath(V v) can than simply return getPath(V u, V v)

d-michail replied:
Now it is just a wrapper around the distance and predecessor matrices of Floyd-Warshall.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ListSingleSourcePaths.java:
```diff
@@ -0,0 +1,116 @@
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ */
+class ListSingleSourcePaths<V, E>
```
Let's make this class public

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/TreeSingleSourcePaths.java:
```diff
@@ -0,0 +1,137 @@
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ */
+class TreeSingleSourcePaths<V, E>
```
Let's make this public

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/TreeSingleSourcePaths.java:
```diff
@@ -0,0 +1,137 @@
+                break;
+            }
+            edgeList.addFirst(e);
+            weight += g.getEdgeWeight(e);
```
Although this is correct, consider changing this to: weight += p.getFirst(); this would even be marginally faster

d-michail replied:
This does not work, as `p.getFirst()` contains the whole path weight.

jkinable commented on jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/DefaultAllPairsPathsTest.java:
```diff
@@ -0,0 +1,78 @@
+
+        for (int i = 1; i < n; i++) {
+            for (int j = 1; j < n; j++) {
+                assertEquals(
```
It would be better if GraphPath has an equals methods which checks the edges, the source and start vertex, and the weight. So you can just check graphPath1.equals(graphPath2)

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BaseShortestPathAlgorithm.java:
```diff
@@ -0,0 +1,103 @@
+                graph, source, sink, Collections.singletonList(source), Collections.emptyList(),
+                0d);
+        } else {
+            return new GraphWalk<>(
```
If we do not return null when there's no path from u to v, then we need to update the descr of GraphPath to inform the user how to verify from the returned path that u and v are in different components?
GraphPath p=alg.getPath(u,v);
if(p==null) //v cannot be reached from u. 

This seems cleaner than:
if(p.getWeight== infinity && p.getVertices.isEmpty)  //v cannot be reached from u.

d-michail replied:
I took back the change and used everywhere `null` when a path does not exist.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DijkstraClosestFirstIterator.java:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    /**
+     * Return the paths computed by this iterator. Only the paths to vertices which are already
```
We should mention somewhere that using Dijkstra as a SingleSourcePaths algorithm is more efficient than repeatedly invoking getPath(u,v) for the same u but different v. FloydWarshall for example doesn't have this advantage.

d-michail replied:
I added a comment in class `DijkstraShortestPath`.

jkinable commented:
Alright, reviewing this monster pull request was quite a bit of work. Some very good changes are included in this pull request!

@jsichi since this is a rather substantial pull request which affects all shortest path algorithms, including some that you designed. It affects the way these algorithms are invoked and results are returned. Could you have a look at the Interface changes and see whether there are specific things you like/don't like/that are missing? Specifically the class ShortestPathAlgorithm, its subinterface and their default implementation classes.

jkinable commented on jgrapht-core/src/test/java/org/jgrapht/perf/shortestpath/DijkstraShortestPathPerformanceTest.java:
```diff
@@ -0,0 +1,286 @@
+
+            generator.generateGraph(graph, new IntegerVertexFactory(), null);
+
+            for (DefaultWeightedEdge e : weightedGraph.edgeSet()) {
```
In one of the tests, you run ALTBenchmark. Is the graph you constructed here feasible for this Admissible Heuristic? Shouldn't the graph's weights obey the triangle inequality?

d-michail replied:
I have added a comment in `ALTAdmissibleHeuristic` that explains that the heuristic does not require that the edge weights satisfy the triangle inequality. 

The method depends on the triangle inequality with respect to the shortest path distances in the graph, not an embedding in Euclidean space or some other metric, which need not be present.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/ALTAdmissibleHeuristic.java:
```diff
@@ -0,0 +1,168 @@
+ * requirement is O(n) per landmark where n is the number of vertices of the graph.
+ * 
+ * <p>
+ * The method generally abbreviated as ALT (from A*, Landmarks and Triangle inequality) is described
```
Could you add a brief description of the method and how it works here (just a few lines summarizing the approach)?

d-michail replied:
done

jsichi commented:
Yeah, I'll take a look soon.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/PathValidator.java:
```diff
@@ -27,8 +27,9 @@
  * @author Assaf Mizrachi
  * @since July, 21, 2016
- *
+ * @deprecated Use {@link org.jgrapht.alg.shortestpath.PathValidator} instead.
```
Assuming we are going to resolve https://github.com/jgrapht/jgrapht/pull/323, shouldn't this be moving to alg.interfaces instead of alg.shortestpath?  It's similar to AStarAdmissibleHeuristic (an interface which the caller of the algorithm implements in order to influence the algorithm's behavior).

d-michail replied:
Indeed. I have not done so, in order to avoid a need of rebasing in the other pull-request. 
 - If the other is merged first, I will rebase this one. 
 - Otherwise, I will submit a new pull-request after the other is merged.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DijkstraShortestPath.java:
```diff
@@ -0,0 +1,111 @@
+
+        return it.getPaths();
+    }
+
```
Same goes for BellmanFord.  If we do bring them back, another possible convention is a package-level utility class (like java.util.Arrays etc).  Whatever we choose will establish a pattern that will need to be followed consistently.

d-michail replied:
I reintroduced the method in each class. It is difficult (naming-wise) to have a package-level utility class since different algorithms make different assumptions.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/ShortestPathAlgorithm.java:
```diff
@@ -0,0 +1,137 @@
+ * 
+ * @author Dimitrios Michail
+ */
+public interface ShortestPathAlgorithm<V, E>
```
done

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/ShortestPathAlgorithm.java:
```diff
@@ -0,0 +1,137 @@
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     */
+    interface AllPairsPaths<V, E>
```
I removed the interface.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/FloydWarshallShortestPaths.java:
```diff
@@ -0,0 +1,437 @@
+     *
+     * @return shortest distance between a and b
+     */
+    public double shortestDistance(V a, V b)
```
done

d-michail commented:
Finally done with the requested changes. Added bonuses, a bug fix in A* and a bug fix in the Fibonacci heap!!

jkinable commented:
hm I can't view the changes made after my last review. Did you squash them together again? :(

d-michail commented:
I did not squash them, but had to rebase due to the merging of the A* bug pull-request. This probably outdated some of the comments. 

However, I can still view all comments, some of them as 'outdated'. There is an extra button at the top-right of each comment, called `show-outdated`.

Unfortunately, they is no going back after rebasing, since its rewrites the revision history.

jkinable commented:
Alright, looks very clean :+1: Thanks. This is going to be a very nice release.

Each time you commit new code, old comments almost always show up as "outdated". That's not problematic. Unfortunately, apparently when you rebase, the wonderful button "view changes" (button next to a review) stops working. This button shows a diff between the last committed version, and the version committed at the time of the review. Certainly for very long reviews like this, that feature is great. 
Now I just clicked on every commit submitted after I reviewed the code to see what changed. That worked, just a little slower :).

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/util/FibonacciHeap.java:
```diff
@@ -254,6 +254,13 @@ public void insert(FibonacciHeapNode<T> node, double key)
+            z.right = z;
+            z.degree = 0;
+            z.child = null;
+            z.mark = false;
```
Yep, parent is already null, because `deleteMin()` always removes a node which is the root of a heap-ordered tree.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/FloydWarshallShortestPaths.java:
```diff
@@ -421,15 +386,55 @@ private void populateLastHopMatrix()
+            int v_b = vertexIndices.get(sink);
+
+            if (backtrace[v_a][v_b] == null) { // No path exists
+                return createEmptyPath(source, sink);
```
Now `createEmptyPath()` returns null if source != sink or a valid `GraphPath` if source == sink.

jkinable commented:
hm, I wanted to merge this pull request but I noticed that there's a small merge conflict caused by the pull request which modified the FibonacciHeap code. Could you fix that? I'll merge this one immediately after to prevent any further conflicts.

d-michail commented:
Conflict fixed

jkinable commented:
great work! Some very nice improvements towards the consistency of the library. Thanks!
